wasm.rkt

```diff
@@ -60,14 +63,92 @@
 ;; Delta
 
 (define-metafunction WASM-eval
-  delta : op v ... -> v
+  delta : op v ... -> v ∨ trap
+  ;;; BINOPS
   [(delta add (const c_0) (const c_1)) (const ,(+ (term c_0) (term c_1)))]
+  [(delta sub (const c_0) (const c_1)) (const ,(- (term c_0) (term c_1)))]
+  [(delta mul (const c_0) (const c_1)) (const ,(* (term c_0) (term c_1)))]
+  [(delta div (const c_0) (const c_1)) ,(if (= (term c_1) 0)
+                                            (term trap)
+                                            (term (const ,(/ (term c_0) (term c_1)))))]
+  [(delta rem (const c_0) (const c_1)) ,(if (= (term c_1) 0)
+                                            (term trap)
+                                            (term (const ,(remainder (term c_0) (term c_1)))))]
+  [(delta and (const c_0) (const c_1)) (const ,(bitwise-and (term c_0) (term c_1)))]
+  [(delta or (const c_0) (const c_1)) (const ,(bitwise-ior (term c_0) (term c_1)))]
+  [(delta xor (const c_0) (const c_1)) (const ,(bitwise-xor (term c_0) (term c_1)))]
+  ; Note that the shift and rotate ops only make sense in the context of a fixed bit width.
+  ; Here, we assume that all operands are 32 bit.
+  [(delta shl (const c_0) (const c_1))
+   (const ,(bitwise-and #xFFFFFFFF
+                        (arithmetic-shift (term c_0) (term c_1))))]
+  [(delta shr (const c_0) (const c_1))
+   (const ,(arithmetic-shift (term c_0) (- (term c_1))))]
+  [(delta rotl (const c_0) (const c_1))
+   (const ,(if (< (term c_1) 0)
```

Comment:
This should be logically equivalent in all cases for fixed bit width

Reply:
Could you explain a bit more? (pun intended)

Reply:
lol, so a left shift of -1 would be a left shift of (-1 mod 32 = 31) according to the spec, which is the same as a right shift of 1.

Smaller example using 8 bits:
```
0b00000100 << -1
0b00000100 << 7
= 0b00000010

0b00000100 << -1
0b00000100 >> 1
= 0b00000010
```

Reply:
I should clarify this only makes sense with rotations not just shifts so pretend << and >> are rotation operators

Reply:
That's very interesting, thanks for clarifying. For increased readability, can we use the "modulo" function in racket instead of "remainder" so we don't have to do this magic?

Reply:
kk